streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -504,88 +438,66 @@ public void update(final Set<TopicPartition> topicPartitions, final Map<String,
         }
 
-        partitionGroup.close();
+        partitionGroup.clear();
         closeTaskSensor.record();
 
         transitionTo(State.CLOSED);
 
         log.info("Closed clean and recycled state");
     }
 
-    /**
-     * <pre>
-     * the following order must be followed:
-     *  1. first close topology to make sure all cached records in the topology are processed
-     *  2. then flush the state, send any left changelog records
-     *  3. then flush the record collector
-     * </pre>
-     *
-     * @param clean    shut down cleanly (ie, incl. flush) if {@code true} --
-     *                 otherwise, just close open resources
-     * @throws TaskMigratedException if the task producer got fenced (EOS)
-     */
-    private void prepareClose(final boolean clean) {
-        // Reset any previously scheduled checkpoint.
-        checkpoint = null;
-
+    private void maybeScheduleCheckpoint() {
         switch (state()) {
-            case CREATED:
-                // the task is created and not initialized, just re-write the checkpoint file
-                scheduleCheckpoint(emptyMap());
+            case RESTORING:
+                this.checkpoint = checkpointableOffsets();
+
                 break;
 
             case RUNNING:
-                closeTopology(clean);
-
-                if (clean) {
-                    stateMgr.flush();
-                    recordCollector.flush();
-                    scheduleCheckpoint(checkpointableOffsets());
-                } else {
-                    executeAndMaybeSwallow(false, stateMgr::flush, "state manager flush", log);
+                if (!eosEnabled) {
+                    this.checkpoint = checkpointableOffsets();
                 }
 
                 break;
 
-            case RESTORING:
-                executeAndMaybeSwallow(clean, stateMgr::flush, "state manager flush", log);
-                scheduleCheckpoint(emptyMap());
+            case SUSPENDED:
+                this.checkpoint = checkpointableOffsets();
 
                 break;
 
-            case SUSPENDED:
+            case CREATED:
             case CLOSED:
-                // not need to checkpoint, since when suspending we've already committed the state
-                break;
+                throw new IllegalStateException("Illegal state " + state() + " while scheduling checkpoint for active task " + id);
 
             default:
-                throw new IllegalStateException("Unknown state " + state() + " while prepare closing active task " + id);
+                throw new IllegalStateException("Unknown state " + state() + " while scheduling checkpoint for active task " + id);
         }
     }
 
-    private void scheduleCheckpoint(final Map<TopicPartition, Long> checkpoint) {
-        this.checkpoint = checkpoint;
-    }
-
     private void writeCheckpointIfNeed() {
+        if (commitNeeded) {
```

Comment:
Why we need this check?

Also nit: how about `maybeWriteCheckpoint` to align with `maybeScheduleCheckpoint`.

Reply:
Why do we need any check? To avoid bugs :)

If like the different names, because "maybe" indicated that the method makes a decision, while "ifNeeded" implies that the methods executed a decision that was already made? At least my personal interpretation?

Reply:
@mjsax  By `should only be written if no commit is needed` do you mean `...if a commit was just completed`? 
Doesn't this break `closeAndRecycleState` and `closeDirty` (since we tend to call `closeDirty` after the commit failed and would not have set `commitNeeded = false`)

Reply:
Maybe I'm thinking of standby tasks (ie we only skip checkpointing for recycled standbys). For active tasks, we should probably commit them before recycling right? Or is it ok to skip committing altogether 🤔

Reply:
Just saw this in the `SmokeTestDriverIntegrationTest#shouldWorkWithRebalance` -- not sure if it merits a separate ticket or can just be fixed together with https://issues.apache.org/jira/browse/KAFKA-10150 ?
```
Caused by: java.lang.IllegalStateException: A checkpoint should only be written if no commit is needed.
	at org.apache.kafka.streams.processor.internals.StreamTask.writeCheckpointIfNeed(StreamTask.java:534)
	at org.apache.kafka.streams.processor.internals.StreamTask.closeAndRecycleState(StreamTask.java:482)
	at org.apache.kafka.streams.processor.internals.StandbyTaskCreator.createStandbyTaskFromActive(StandbyTaskCreator.java:115)
	at org.apache.kafka.streams.processor.internals.TaskManager.handleAssignment(TaskManager.java:288)
```

Reply:
Nice catch. Let's just fix it along with 10150?

Reply:
Too late, I already created a ticket for it 🙂 But after starting to work on it, I agree, they should be addressed in one PR